File: src/main/java/com/koddy/server/member/domain/model/response/LanguageResponse.java
package com.koddy.server.member.domain.model.response;

import com.koddy.server.member.domain.model.Language;

import java.util.List;

public record LanguageResponse(
        String main,
        List<String> sub
) {
    public static LanguageResponse of(final List<Language> languages) {
        return new LanguageResponse(getMainLanguage(languages), getSubLanguages(languages));
    }

    private static String getMainLanguage(final List<Language> languages) {
        return languages.stream()
                .filter(it -> it.getType() == Language.Type.MAIN)
                .findFirst()
                .orElseThrow(RuntimeException::new)
                .getCategory()
                .getCode();
    }

    private static List<String> getSubLanguages(final List<Language> languages) {
        return languages.stream()
                .filter(it -> it.getType() == Language.Type.SUB)
                .map(it -> it.getCategory().getCode())
                .toList();
    }
}

File: src/main/java/com/koddy/server/member/domain/model/response/LanguageResponse.kt
package com.koddy.server.member.domain.model.response

import com.koddy.server.member.domain.model.Language

data class LanguageResponse(
    val main: String,
    val sub: List<String> = emptyList(),
) {
    companion object {
        fun of(languages: List<Language>): LanguageResponse {
            return LanguageResponse(
                main = getMainLanguage(languages),
                sub = getSubLanguages(languages),
            )
        }

        private fun getMainLanguage(languages: List<Language>): String {
            return languages
                .first { it.type == Language.Type.MAIN }
                .category
                .code
        }

        private fun getSubLanguages(languages: List<Language>): List<String> {
            return languages
                .filter { it.type == Language.Type.SUB }
                .map { it.category.code }
                .toList()
        }
    }
}
